### DISCUSSION TITLE
Resize image without changing bits per pixel (Magick++)

### DISCUSSION THREAD

**Discussion by MrAnderson1971** [ImageMagick/ImageMagick#6373] [Help] [answered]
### ImageMagick version

7.1.1-Q8

### Operating system, version and so on

Windows 10 (x86)

### Description

Whenever I resize an image up or down, there's a chance the bits per pixel might change, which I don't want.

I'm testing it on 1-bit monochrome, 8-bit grayscale, 8-bit color, 16-bit grayscale, and 24-bit grayscale.
And if it matters, I'm saving them as BMP.

Thank you.

**Comment by fmw42** [comment]
I believe Imagemagick will keep the depth, if the operation does not change the data, such as when you introduce new colors by resizing. I suspect you will need to add -depth X in your command to keep that from happening, where X is your desired output depth.

**Reply by MrAnderson1971** [reply]
Okay, I've looked into it. It's not the depth changing that's causing the bits per pixel to change. Rather the number of channels sometimes changes when resizing.

**Reply by fmw42** [reply]
The number of channels should not change with a resize.  Please post an example and your command where that happens.  As I said before, when resizing you change the number of colors. The number may go more than 255 if you start with 8-bits. This would cause the result to be 16-bits if on Q16.

**Comment by snibgo** [comment]
> Rather the number of channels sometimes changes when resizing.

Please show your Magick++ code, and sample input and output images. Otherwise, we can only guess.

**Reply by MrAnderson1971** [reply, marked as answer]
[blissgrayscaleinput.zip](https://github.com/ImageMagick/ImageMagick/files/11602773/blissgrayscaleinput.zip)
`      image.read(blob);      
      size_t channels1 = image.channels();
      Magick::Geometry newSize(newX, newY);
      newSize.aspect(true);
      image.filterType(Magick::FilterType::CatromFilter); 
      image.resize(newSize); 
      size_t channels2 = image.channels();`
channels1 = 4 
channels2 = 3

But I was able to fix it by checking if channels * depth was equal to the original bits per pixel, and if not, quantizing it manually.

**Comment by fmw42** [comment]
your input number of channels is wrong.

identify -verbose shows only 3 equal input channels (grayscale image)

```
 Channel statistics:
    Pixels: 480000
    Red:
      min: 5  (0.0196078)
      max: 254 (0.996078)
      mean: 129.229 (0.506781)
      standard deviation: 48.3209 (0.189494)
      kurtosis: -0.464075
      skewness: 0.310267
      entropy: 0.932334
    Green:
      min: 5  (0.0196078)
      max: 254 (0.996078)
      mean: 129.229 (0.506781)
      standard deviation: 48.3209 (0.189494)
      kurtosis: -0.464075
      skewness: 0.310267
      entropy: 0.932334
    Blue:
      min: 5  (0.0196078)
      max: 254 (0.996078)
      mean: 129.229 (0.506781)
      standard deviation: 48.3209 (0.189494)
      kurtosis: -0.464075
      skewness: 0.310267
      entropy: 0.932334

```
